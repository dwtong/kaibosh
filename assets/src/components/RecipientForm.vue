<template>
  <form @submit.prevent="submit">
    <div class="box">
      <h1 class="title">Organisation Details</h1>

      <b-field
        label="Organisation name"
        :type="{ 'is-danger': errors.has('name') }"
        :message="errors.first('name')"
        class="form-field"
      >
        <b-input v-model="recipient.name" name="name" v-validate="'required'" />
      </b-field>

      <b-field
        label="Kaibosh base"
        :type="{ 'is-danger': errors.has('base') }"
        :message="errors.first('base')"
        class="form-field"
      >
        <b-select
          v-model="recipient.base_id"
          name="base"
          placeholder="Select an option"
          v-validate="'required'"
        >
          <option v-for="base in bases.data" :key="base.id" :value="base.id">{{
            base.name
          }}</option>
        </b-select>
      </b-field>
    </div>

    <div class="box">
      <h1 class="title">Primary Contact Details</h1>

      <b-field
        label="Contact name"
        :type="{ 'is-danger': errors.has('contact-name') }"
        :message="errors.first('contact-name')"
      >
        <b-input
          v-model="recipient.primary_contact.name"
          data-vv-as="contact name"
          name="contact-name"
        />
      </b-field>

      <b-field
        label="Contact email"
        :type="{ 'is-danger': errors.has('primary_contact.email') }"
        :message="errors.first('primary_contact.email')"
      >
        <b-input
          data-vv-as="contact email"
          v-model="recipient.primary_contact.email"
          name="contact-email"
        />
      </b-field>

      <b-field label="Contact landline">
        <b-input
          v-model="recipient.primary_contact.phone_landline"
          name="contact-landline"
        />
      </b-field>

      <b-field label="Contact mobile">
        <b-input v-model="recipient.primary_contact.phone_mobile" name="name" />
      </b-field>
    </div>

    <button type="submit" class="button is-primary is-pulled-right">
      Save Recipient
    </button>
  </form>
</template>

<script>
import { mapActions, mapState } from "vuex";

export default {
  async created() {
    await this.getBases();
  },

  computed: {
    ...mapState("bases", ["bases"])
  },

  inject: ["$validator"],

  methods: {
    ...mapActions("bases", ["getBases"]),
    async submit() {
      this.$emit("submit", this.recipient);
    }
  },

  props: {
    recipient: {
      type: Object,
      default() {
        return {
          name: null,
          base_id: null,
          primary_contact: {
            name: null,
            email: null,
            phone_landline: null,
            phone_mobile: null
          }
        };
      }
    }
  }
};
</script>
