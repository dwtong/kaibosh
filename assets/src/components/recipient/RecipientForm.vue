<template>
  <form @submit.prevent="submit">
    <b-loading :active.sync="isLoading"></b-loading>
    <div class="box">
      <h1 class="title">Organisation Details</h1>

      <InputField name="name" v-model="recipient.name" required="true" />
      <BaseSelect v-model="recipient.baseId" label="Kaibosh base" required="true" />
      <AddressField v-model="recipient.physicalAddress" label="Physical address" />
      <DateField name="start date" v-model="startedAt" />
      <InputField name="description" v-model="recipient.description" type="textarea" />
    </div>

    <div class="box">
      <h1 class="title">Primary Contact Details</h1>

      <InputField name="contact-name" v-model="contact.name" />
      <InputField name="contact-email" v-model="contact.email" />
      <InputField name="contact-landline" v-model="contact.phoneLandline" />
      <InputField name="contact-mobile" v-model="contact.phoneMobile" />
    </div>
    <button type="submit" class="button is-primary is-pulled-right">
      Save Recipient
    </button>
  </form>
</template>

<script lang="ts">
import Vue from "vue";
import { Component, Prop } from "vue-property-decorator";
import AddressField from "@/components/ui/AddressField.vue";
import BaseSelect from "@/components/ui/BaseSelect.vue";
import DateField from "@/components/ui/DateField.vue";
import InputField from "@/components/ui/InputField.vue";
import { IContact, IRecipient } from "@/types";
import { ActiveRecipientModule } from "@/store/modules/active-recipient";
import toast from "@/helpers/toast";

@Component({
  components: { AddressField, BaseSelect, DateField, InputField },
  // eslint-disable-next-line @typescript-eslint/camelcase
  $_veeValidate: { validator: "new" }
})
export default class RecipientForm extends Vue {
  @Prop(String) readonly recipientId!: string;

  recipient: IRecipient = {
    name: ""
  };

  contact: IContact = {
    name: "",
    email: "",
    phoneLandline: "",
    phoneMobile: ""
  };

  startedAt: Date = new Date();
  isLoading = true;

  async submit() {
    const formIsValid = await this.$validator.validateAll();

    if (formIsValid) {
      this.saveRecipient();
    } else {
      toast.error("Failed to save recipient.", true);
    }
  }

  async created() {
    if (this.recipientId) {
      await ActiveRecipientModule.fetchRecipient(this.recipientId);
      this.recipient = ActiveRecipientModule.details;

      if (ActiveRecipientModule.details.contact) {
        this.contact = { ...ActiveRecipientModule.details.contact };
      }

      if (ActiveRecipientModule.details.startedAt) {
        this.startedAt = ActiveRecipientModule.details.startedAt;
      }
    }

    this.isLoading = false;
  }

  async saveRecipient() {
    const params = {
      recipient: {
        ...this.recipient,
        contact: this.contact
      }
    };

    if (this.startedAt) {
      params.recipient.startedAt = this.startedAt;
    }

    if (this.recipientId) {
      await ActiveRecipientModule.updateRecipient(params);
      this.$router.push(`/recipients/${this.recipientId}`);
      toast.success("Recipient updated.");
    } else {
      await ActiveRecipientModule.createRecipient(params);

      if (ActiveRecipientModule.details.id) {
        this.$router.push(`/recipients/${ActiveRecipientModule.details.id}`);
        toast.success("Recipient created.");
      }
    }
  }
}
</script>
