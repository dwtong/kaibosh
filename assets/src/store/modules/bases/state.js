export default {
  loading: false,
  errors: [],
  list: [],
  sessionSlots: [],
  foodCategories: []
};
