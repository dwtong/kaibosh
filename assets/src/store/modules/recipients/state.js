export default {
  errors: null,
  loading: false,

  list: [],

  details: {
    name: "",
    id: null,
    status: "",
    primary_contact: {
      name: "",
      email: "",
      mobile: "",
      landline: ""
    },
    base: {
      name: "",
      id: null
    }
  },

  scheduledSessions: [],

  filters: {
    base_id: 0,
    name: "",
    status: [
      { label: "Active", name: "active", enabled: false },
      { label: "Pending", name: "pending", enabled: false },
      { label: "On Hold", name: "on_hold", enabled: false },
      { label: "Archived", name: "archived", enabled: false }
    ]
  }
};
