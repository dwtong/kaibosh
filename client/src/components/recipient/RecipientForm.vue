<script setup lang="ts">
import { computed, ref } from "vue"
import { useForm } from "vee-validate"
import { object, string, date } from "zod"
import { toTypedSchema } from "@vee-validate/zod"
import SubmitButton from "../ui/SubmitButton.vue"
import ValidatedInput from "../ui/ValidatedInput.vue"
import ValidatedBaseSelect from "../ui/ValidatedBaseSelect.vue"
import ValidatedDate from "../ui/ValidatedDate.vue"
import ValidatedAddress from "../ui/ValidatedAddress.vue"
import type { Recipient } from "@/api/recipients"

const props = defineProps<{
  recipient?: Recipient
  onSubmit: (recipient: Partial<Recipient>) => Promise<void>
}>()

console.log(props.recipient?.physicalAddress)

const defaults = {
  name: props.recipient?.name || "",
  baseId: props.recipient?.baseId.toString() || "",
  description: props.recipient?.description || "",
  contactName: props.recipient?.contact.name || "",
  contactEmail: props.recipient?.contact.email || "",
  contactPhoneLandline: props.recipient?.contact.phoneLandline || "",
  contactPhoneMobile: props.recipient?.contact.phoneMobile || "",
  physicalAddress: props.recipient?.physicalAddress || "",
  startDate: props.recipient?.startedAt
    ? new Date(props.recipient?.startedAt)
    : null,
}

const validationSchema = toTypedSchema(
  object({
    name: string().nonempty("Name is required").default(defaults.name),
    baseId: string().nonempty("Base is required").default(defaults.baseId),
    description: string().optional().default(defaults.description),
    startDate: date().optional().nullable().default(defaults.startDate),
    contactName: string().optional().default(defaults.contactName),
    contactEmail: string()
      .email()
      .optional()
      .nullable()
      .default(defaults.contactEmail || null),
    contactPhoneLandline: string()
      .optional()
      .default(defaults.contactPhoneLandline),
    contactPhoneMobile: string()
      .optional()
      .default(defaults.contactPhoneMobile),
    physicalAddress: string().optional().default(defaults.physicalAddress),
  }),
)
const { handleSubmit, submitCount } = useForm({
  validationSchema,
})
const showErrors = computed(() => submitCount.value > 0)
const isSubmitting = ref(false)
const submit = handleSubmit(async (recipientParams) => {
  isSubmitting.value = true
  const {
    baseId,
    contactName,
    contactEmail,
    contactPhoneMobile,
    contactPhoneLandline,
    startDate,
    ...recipientDetails
  } = recipientParams
  const recipient: Partial<Recipient> = {
    ...recipientDetails,
    baseId: parseInt(baseId),
    startedAt: startDate,
    contact: {
      name: contactName || "",
      email: contactEmail || "",
      phoneMobile: contactPhoneMobile || "",
      phoneLandline: contactPhoneLandline || "",
    },
  }
  await props.onSubmit(recipient)
  isSubmitting.value = false
})
</script>

<template>
  <form
    class="form"
    :validation-schema="validationSchema"
    @submit.prevent="submit"
  >
    <div class="box">
      <h1 class="title">Organisation Details</h1>
      <ValidatedInput name="name" type="text" :show-error="showErrors" />
      <ValidatedBaseSelect
        name="baseId"
        label="kaibosh base"
        placeholder="Select a base..."
        :show-error="showErrors"
      />
      <ValidatedAddress
        name="physicalAddress"
        type="text"
        label="physical address"
        :value="defaults.physicalAddress"
        :show-error="showErrors"
      />
      <ValidatedDate
        name="startDate"
        label="start date"
        :show-error="showErrors"
        placeholder="Click to select..."
      />
      <ValidatedInput
        name="description"
        type="textarea"
        :show-error="showErrors"
      />
    </div>

    <div class="box">
      <h1 class="title">Primary Contact Details</h1>

      <ValidatedInput
        name="contactName"
        label="contact name"
        type="text"
        :show-error="showErrors"
      />
      <ValidatedInput
        name="contactEmail"
        label="contact email"
        type="text"
        :show-error="showErrors"
      />
      <ValidatedInput
        name="contactPhoneLandline"
        label="contact phone (landline)"
        type="text"
        :show-error="showErrors"
      />
      <ValidatedInput
        name="contactPhoneMobile"
        label="contact phone (mobile)"
        type="text"
        :show-error="showErrors"
      />
      <!-- <ValidatedInput v-model="recipientDetails.contact.name" label="contact name" /> -->
      <!-- <ValidatedInput v-model="recipientDetails.contact.email" label="contact email" :rules="{ email: true }" /> -->
      <!-- <ValidatedInput v-model="recipientDetails.contact.phoneLandline" label="contact landline" /> -->
      <!-- <ValidatedInput v-model="recipientDetails.contact.phoneMobile" label="contact mobile" /> -->
    </div>
    <SubmitButton
      :is-submitting="isSubmitting"
      class="button is-primary is-pulled-right"
      label="Save Recipient"
    />
  </form>
</template>
