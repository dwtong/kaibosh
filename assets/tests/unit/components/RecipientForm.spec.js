import { mount, shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import Buefy from "buefy";
import VeeValidate from "vee-validate";
import RecipientForm from "@/components/RecipientForm.vue";
import state from "@/store/modules/bases/state";

const localVue = createLocalVue();
const bases = [{ id: 1, name: "Base 1" }, { id: 2, name: "Base 2" }];

localVue.use(Vuex);
localVue.use(Buefy);
localVue.use(VeeValidate);

describe("RecipientForm.vue", () => {
  let actions;
  let store;
  let provide;

  beforeEach(() => {
    actions = {
      getBases: jest.fn()
    };

    provide = {
      $validator: new VeeValidate.Validator()
    };

    store = new Vuex.Store({
      modules: {
        bases: {
          namespaced: true,
          actions,
          state: { ...state, bases: bases }
        }
      }
    });
  });

  describe("component layout", () => {
    it("renders correctly", () => {
      const wrapper = mount(RecipientForm, { store, localVue, provide });
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe("submitting form", () => {
    it("emits submit event", () => {
      const wrapper = shallowMount(RecipientForm, { store, localVue, provide });

      const form = wrapper.find("form");
      form.trigger("submit");
      expect(wrapper.emitted().submit).toBeTruthy();
    });
  });
});
