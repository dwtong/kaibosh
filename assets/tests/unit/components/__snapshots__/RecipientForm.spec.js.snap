// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipientForm.vue component layout renders correctly 1`] = `
<form>
  <div
    class="box"
  >
    <h1
      class="title"
    >
      Organisation Details
    </h1>
     
    <div
      class="field form-field"
    >
      <label
        class="label"
      >
        
            Organisation name
        
      </label>
       
      <div
        class="control is-clearfix"
      >
        <input
          autocomplete="on"
          class="input"
          name="name"
          type="text"
        />
         
        <!---->
         
        <!---->
         
        <!---->
      </div>
       
      <!---->
    </div>
     
    <div
      class="field form-field"
    >
      <label
        class="label"
      >
        
            Kaibosh base
        
      </label>
       
      <div
        class="control"
      >
        <span
          class="select is-empty"
        >
          <select
            name="base"
          >
            <option
              disabled="disabled"
              hidden="hidden"
              selected="selected"
              value=""
            >
              
                    Select an option
                
            </option>
             
          </select>
        </span>
         
        <!---->
      </div>
       
      <!---->
    </div>
  </div>
   
  <div
    class="box"
  >
    <h1
      class="title"
    >
      Primary Contact Details
    </h1>
     
    <div
      class="field"
    >
      <label
        class="label"
      >
        
            Contact name
        
      </label>
       
      <div
        class="control is-clearfix"
      >
        <input
          autocomplete="on"
          class="input"
          data-vv-as="contact name"
          name="contact-name"
          type="text"
        />
         
        <!---->
         
        <!---->
         
        <!---->
      </div>
       
      <!---->
    </div>
     
    <div
      class="field"
    >
      <label
        class="label"
      >
        
            Contact email
        
      </label>
       
      <div
        class="control is-clearfix"
      >
        <input
          autocomplete="on"
          class="input"
          data-vv-as="contact email"
          name="contact-email"
          type="text"
        />
         
        <!---->
         
        <!---->
         
        <!---->
      </div>
       
      <!---->
    </div>
     
    <div
      class="field"
    >
      <label
        class="label"
      >
        
            Contact landline
        
      </label>
       
      <div
        class="control is-clearfix"
      >
        <input
          autocomplete="on"
          class="input"
          name="contact-landline"
          type="text"
        />
         
        <!---->
         
        <!---->
         
        <!---->
      </div>
       
      <!---->
    </div>
     
    <div
      class="field"
    >
      <label
        class="label"
      >
        
            Contact mobile
        
      </label>
       
      <div
        class="control is-clearfix"
      >
        <input
          autocomplete="on"
          class="input"
          name="name"
          type="text"
        />
         
        <!---->
         
        <!---->
         
        <!---->
      </div>
       
      <!---->
    </div>
  </div>
   
  <button
    class="button is-primary is-pulled-right"
    type="submit"
  >
    
    Save Recipient
  
  </button>
</form>
`;
